src/core/unit-serialize.c

```diff
@@ -89,13 +89,21 @@ static const char *const io_accounting_metric_field_last[_CGROUP_IO_ACCOUNTING_M
         [CGROUP_IO_WRITE_OPERATIONS] = "io-accounting-write-operations-last",
 };
 
-int unit_serialize(Unit *u, FILE *f, FDSet *fds, bool serialize_jobs) {
+int unit_serialize(Unit *u, FILE *f, FDSet *fds, bool switching_root) {
         int r;
 
         assert(u);
         assert(f);
         assert(fds);
 
+        if (switching_root && IN_SET(u->type, UNIT_MOUNT, UNIT_AUTOMOUNT)) {
+                /* In the new root, paths for mounts and automounts will be different, so it doesn't make
+                 * much sense to serialize things. API file systems will be moved to the new root, but we
+                 * don't have mount points for those. */
+                log_unit_debug(u, "not serializing before switch-root");
+                return 0;
+        }
```

Comment:
hmm, i think it would be much nicer to have a boolean for this in UnitVTable, which is how we usually do that. i.e.

```
if (switching_root && !UNIT_VTABLE(u)->serialize_on_switch_root) {
        log_unit_debug(u, "Unit type excluded from serialization during switch-root.");
        return 0;
}
```

Reply:
Would it? I considered adding that, but then the logic would be spread out over 4 places:
- unit.h to declare the field
- automount.c to set it
- mount.c to set it
- and here to check it

I like encapsulation and everything, but this is all in one codebase, so I like to have put the check here, with a nice comment, instead of splitting it up so much.

Reply:
i think having per-unit-type code in unit.c and manager.c kinda sucks. Ideally we'd have those files free from any form of unit type checking...

Reply:
Updated to put the info in VTable.